File: IsotopeID/MatlabAndBef/read_bef.m
% function read_bef(filename)
% filedir = 'I:\M400\20240515-200531_0cm';
% filedir = 'I:\M400\20240605-165232_20240605_cs137_alone_30_min';
filedir = 'C:\Users\Jacob\Desktop\m400';

fid = fopen(strcat(filedir, '\AllEventsEu152.bef'));
listmodefile = fopen(strcat(filedir, '\AllEvents_listMode.txt'), 'w');

if fid < 0
    return;
end

if listmodefile < 0
    return;
end

fseek(fid, 0, 'eof');
fsize = ftell(fid);
fseek(fid, 0, 'bof');

edges_e = 0:1000;
hist2debef = zeros(length(edges_e), 5);

totevt = 0;
bSkipWrongData = 0;
tic;
while (~feof(fid))
    [cdatatype, nread] = fread(fid, 1, "int8");
    if nread <= 0; break; end

    if cdatatype <= 0 && bSkipWrongData == 0
        if cdatatype == 0
            nFormatVersion = fread(fid, 1, "uint32");
            strIntro = fread(fid, 20, 'char');
            
            nLen = fread(fid, 1, 'uint16');
            strFilename = fread(fid, nLen, 'char');

            fprintf('Format version = %d\n', nFormatVersion);
            fprintf('  =>: %s\n', strIntro);
            fprintf('  =>: %s\n', strFilename);

            nTimeStart = fread(fid, 1, 'uint32');
            nTimeStartFraction = fread(fid, 1, 'uint32');
            nFPGAStart = fread(fid, 1, 'uint32');
            nFPGAStartFraction = fread(fid, 1, 'uint32');
            cBEFEnergyOnly = fread(fid, 1, 'uint8');
            if nFormatVersion >=8
                cIndMode = fread(fid, 1, 'char');
            end
            nAdditionalHeaderLength = fread(fid, 1, 'uint32');

            fprintf('Computer Time: %d - %d\n', nTimeStart, nTimeStartFraction);
            fprintf('FPGA Time: %d - %d\n', nFPGAStart, nFPGAStartFraction);

        elseif cdatatype == -4
            nLen = fread(fid, 1, 'uint16');
            strDeviceID = fread(fid, nLen, 'uint8');
            fprintf("Device ID: %s\n", strDeviceID);

        elseif cdatatype == -100
            dwUnusded = fread(fid, 3, 'uint32');
        else
            fprintf("Unknown data type %d\n", cdatatype);
            dwUnusded0 = fread(fid, 1, 'uint32');
            dwUnusded1 = fread(fid, 1, 'uint32');
            dwUnusded2 = fread(fid, 1, 'uint32');
            fprintf("Wrong Data:%d %d %d\n", dwUnusded0, dwUnusded1, dwUnusded2);
        end
    else
        if cdatatype > 0 && cdatatype <= 121

            nEnergy = zeros(121, 1);
            nModuleID = zeros(121, 1);
            nX = zeros(121, 1);
            nY = zeros(121, 1);
            nZ = zeros(121, 1);
    
            npix = 1;
            if nFormatVersion<7 && cBEFEnergyOnly==1
                nEnergy(1) = fread(fid, 1, 'uint32');
            else
                npix = cdatatype;
                for ii=1:cdatatype
                    if cBEFEnergyOnly==1 && (nFormatVersion >= 8 || (nFormatVersion == 7 && cIndMode == 1))
                        nModuleID(ii) = fread(fid, 1, 'char');
                    end
    
                    if nFormatVersion <= 5
                        nEnergy(ii) = fread(fid, 1, 'uint16');
                    else
                        nEnergy(ii) = fread(fid, 1, 'uint32');
                    end
                    
                    if cBEFEnergyOnly==0
                        nX(ii) = fread(fid, 1, 'short');
                        nY(ii) = fread(fid, 1, 'short');
                        nZ(ii) = fread(fid, 1, 'short');
                    end
                end
            end

            nEventTime = fread(fid, 1, 'uint32');
            dwLiveTime = fread(fid, 1, 'uint32');
    
            if(nFormatVersion <= 5)
                totEnergy = sum(nEnergy)/10;
            else
                totEnergy = sum(nEnergy)/1000;
            end
             
            % user code start here
            [~, idxe] = max(edges_e > totEnergy);
            
            if npix <= 4
                hist2debef(idxe, npix) = hist2debef(idxe, npix) + 1;
            else
                hist2debef(idxe, 5) = hist2debef(idxe, 5) + 1;
            end

            % user code end here

            for i = 1:npix
                newLine = sprintf('%d %10d %10d %10d %10.2f %10.2f %10.2f\n', int32(nEventTime), int32(npix), int32(nEnergy(i)), int32(nModuleID(i)), nX(i), nY(i), nZ(i));
                fprintf(listmodefile, newLine);
            end

        else
            bSkipWrongData = 1;
        end
    end

    totevt = totevt + 1;
    if mod(totevt, 1e6) == 0
        curtm = toc;
        fpos = ftell(fid);
        fprintf('--: Elasped Time: %.1f seconds, Total Time: %.1f seconds\n', curtm, curtm / fpos * fsize);
    end
end
toc;

fclose all;

save(sprintf('%s\\befspec.mat', filedir), 'hist2debef');

%%
plot(hist2debef)

File: Isotopes location finder/read_bef.py
import os
import struct
import numpy as np
import time
import matplotlib.pyplot as plt
import sys
import csv

def read_bef(filedir):
    fid_path = os.path.join(filedir, 'AllEventsEu152.bef')
    listmodefile_path = os.path.join(filedir, 'AllEvents_listMode.csv')  # Change file extension to CSV

    if not os.path.exists(fid_path):
        print(f"File not found: {fid_path}")
        return

    with open(fid_path, 'rb') as fid, open(listmodefile_path, 'w', newline='') as csvfile:
        edges_e = np.arange(1001)
        hist2debef = np.zeros((len(edges_e), 5))
        
        totevt = 0
        bSkipWrongData = 0
        start_time = time.time()

        # Set up CSV writer
        csv_writer = csv.writer(csvfile)
        # Write header if needed
        csv_writer.writerow(['EventTime', 'NumPixels', 'Energy', 'ModuleID', 'X', 'Y', 'Z'])
        
        fid.seek(0, os.SEEK_END)
        fsize = fid.tell()
        fid.seek(0, os.SEEK_SET)

        while fid.tell() < fsize:
            cdatatype_data = fid.read(1)
            if len(cdatatype_data) == 0:
                break
                
            cdatatype = struct.unpack("b", cdatatype_data)[0]

            if cdatatype <= 0 and bSkipWrongData == 0:
                if cdatatype == 0:
                    try:
                        nFormatVersion = struct.unpack("I", fid.read(4))[0]
                        strIntro = fid.read(20)
                        
                        nLen = struct.unpack("H", fid.read(2))[0]
                        strFilename = fid.read(nLen)
                        
                        print(f'Format version = {nFormatVersion}')
                        print(f'  =>: {strIntro.decode().strip()}')
                        print(f'  =>: {strFilename.decode().strip()}')

                        nTimeStart = struct.unpack("I", fid.read(4))[0]
                        nTimeStartFraction = struct.unpack("I", fid.read(4))[0]
                        nFPGAStart = struct.unpack("I", fid.read(4))[0]
                        nFPGAStartFraction = struct.unpack("I", fid.read(4))[0]
                        cBEFEnergyOnly = struct.unpack("B", fid.read(1))[0]

                        if nFormatVersion >= 8:
                            cIndMode = struct.unpack("c", fid.read(1))[0]

                        nAdditionalHeaderLength = struct.unpack("I", fid.read(4))[0]

                        print(f'Computer Time: {nTimeStart} - {nTimeStartFraction}')
                        print(f'FPGA Time: {nFPGAStart} - {nFPGAStartFraction}')
                    except struct.error as e:
                        print("Error while unpacking header data:", e)
                        break

                elif cdatatype == -4:
                    try:
                        nLen = struct.unpack("H", fid.read(2))[0]
                        strDeviceID = fid.read(nLen)
                        print(f"Device ID: {strDeviceID.decode().strip()}")
                    except struct.error as e:
                        print("Error while unpacking Device ID:", e)
                        break

                elif cdatatype == -100:
                    try:
                        unused_data_bytes = fid.read(12)
                        if len(unused_data_bytes) == 12:  # Ensure we read enough bytes
                            dwUnusded = struct.unpack("3I", unused_data_bytes)
                        else:
                            print("Error: Not enough data for unused data")
                            break
                    except struct.error as e:
                        print("Error while unpacking unused data:", e)
                        break

                else:
                    print(f"Unknown data type {cdatatype}")
                    try:
                        dwUnusded0 = struct.unpack("I", fid.read(4))[0]
                        dwUnusded1 = struct.unpack("I", fid.read(4))[0]
                        dwUnusded2 = struct.unpack("I", fid.read(4))[0]
                        print(f"Wrong Data: {dwUnusded0} {dwUnusded1} {dwUnusded2}")
                    except struct.error as e:
                        print("Error while unpacking wrong data:", e)
                        break

            else:
                if 0 < cdatatype <= 121:
                    nEnergy = np.zeros(121)
                    nModuleID = np.zeros(121)
                    nX = np.zeros(121)
                    nY = np.zeros(121)
                    nZ = np.zeros(121)

                    try:
                        npix = 1
                        if nFormatVersion < 7 and cBEFEnergyOnly == 1:
                            nEnergy[0] = struct.unpack("I", fid.read(4))[0]
                        else:
                            npix = cdatatype
                            for ii in range(cdatatype):
                                if cBEFEnergyOnly == 1 and (nFormatVersion >= 8 or (nFormatVersion == 7 and cIndMode == 1)):
                                    nModuleID[ii] = struct.unpack("c", fid.read(1))[0]
                                    
                                if nFormatVersion <= 5:
                                    nEnergy[ii] = struct.unpack("H", fid.read(2))[0]
                                else:
                                    nEnergy[ii] = struct.unpack("I", fid.read(4))[0]
                        
                                if cBEFEnergyOnly == 0:
                                    nX[ii] = struct.unpack("h", fid.read(2))[0]
                                    nY[ii] = struct.unpack("h", fid.read(2))[0]
                                    nZ[ii] = struct.unpack("h", fid.read(2))[0]

                        nEventTime = struct.unpack("I", fid.read(4))[0]
                        dwLiveTime = struct.unpack("I", fid.read(4))[0]

                    except struct.error as e:
                        print("Error while unpacking event data:", e)
                        break
    
                    totEnergy = np.sum(nEnergy) / (1000 if nFormatVersion > 5 else 10)
             
                    idxe = np.argmax(edges_e > totEnergy)
                    
                    if npix <= 4:
                        hist2debef[idxe, npix] += 1
                    else:
                        hist2debef[idxe, 4] += 1
                    
                    for i in range(npix):
                        row = [int(nEventTime), int(npix), int(nEnergy[i]), int(nModuleID[i]), nX[i], nY[i], nZ[i]]
                        csv_writer.writerow(row)

                else:
                    bSkipWrongData = 1

            totevt += 1
            if totevt % 1e6 == 0:
                curtm = time.time() - start_time
                fpos = fid.tell()
                print(f'--: Elapsed Time: {curtm:.1f} seconds, Total Time: {curtm / fpos * fsize:.1f} seconds')

        np.save(os.path.join(filedir, 'befspec.npy'), hist2debef)

    plt.plot(hist2debef)
    plt.show()

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python read_bef.py <file_directory>")
    else:
        file_directory = sys.argv[1]
        read_bef(file_directory)
